File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/sync/specs/GivenAnInternalStoragePreference/WhenLookingUpTheSyncDrive.java
package com.lasthopesoftware.bluewater.client.stored.library.sync.specs.GivenAnInternalStoragePreference;

import com.lasthopesoftware.bluewater.client.library.access.ILibraryProvider;
import com.lasthopesoftware.bluewater.client.library.repository.LibraryId;
import com.lasthopesoftware.bluewater.client.stored.library.sync.SyncDirectoryLookup;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import com.lasthopesoftware.storage.directories.specs.FakePrivateDirectoryLookup;
import com.lasthopesoftware.storage.directories.specs.FakePublicDirectoryLookup;

import org.junit.BeforeClass;
import org.junit.Test;

import java.io.File;
import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
import static org.mockito.Mockito.mock;

public class WhenLookingUpTheSyncDrive {

	private static File file;

	@BeforeClass
	public static void before() throws ExecutionException, InterruptedException {
		final FakePrivateDirectoryLookup fakePrivateDirectoryLookup = new FakePrivateDirectoryLookup();
		fakePrivateDirectoryLookup.addDirectory("", 1);
		fakePrivateDirectoryLookup.addDirectory("", 2);
		fakePrivateDirectoryLookup.addDirectory("", 3);
		fakePrivateDirectoryLookup.addDirectory("/storage/0/my-private-sd-card", 10);

		final FakePublicDirectoryLookup publicDrives = new FakePublicDirectoryLookup();
		publicDrives.addDirectory("fake-private-path", 12);
		publicDrives.addDirectory("/fake-private-path", 5);

		final SyncDirectoryLookup syncDirectoryLookup = new SyncDirectoryLookup(
			mock(ILibraryProvider.class),
			publicDrives,
			fakePrivateDirectoryLookup);

		file = new FuturePromise<>(
			syncDirectoryLookup.promiseSyncDirectory(
				new LibraryId(1)
				/*new Library()
				.setSyncedFileLocation(Library.SyncedFileLocation.INTERNAL)*/)).get();
	}

	@Test
	public void thenTheDriveIsTheOneWithTheMostSpace() {
		assertThat(file.getPath()).isEqualTo("/storage/0/my-private-sd-card");
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/sync/GivenAnInternalStoragePreference/WhenLookingUpTheSyncDrive.kt
package com.lasthopesoftware.bluewater.client.stored.library.sync.GivenAnInternalStoragePreference

import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.browsing.library.repository.SyncedFileLocation
import com.lasthopesoftware.bluewater.client.browsing.library.settings.LibrarySettings
import com.lasthopesoftware.bluewater.client.browsing.library.settings.StoredMediaCenterConnectionSettings
import com.lasthopesoftware.bluewater.client.stored.library.sync.SyncDirectoryLookup
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.lasthopesoftware.promises.extensions.toPromise
import com.lasthopesoftware.storage.directories.FakePrivateDirectoryLookup
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.junit.jupiter.api.Test

class WhenLookingUpTheSyncDrive {
	private val file by lazy {
		val fakePrivateDirectoryLookup = FakePrivateDirectoryLookup()
		fakePrivateDirectoryLookup.addDirectory("", 1)
		fakePrivateDirectoryLookup.addDirectory("", 2)
		fakePrivateDirectoryLookup.addDirectory("", 3)
		fakePrivateDirectoryLookup.addDirectory("/storage/0/my-private-sd-card", 10)
		val syncDirectoryLookup = SyncDirectoryLookup(
			mockk {
				every { promiseLibrarySettings(LibraryId(1)) } returns LibrarySettings(
					libraryId = LibraryId(1),
					syncedFileLocation = SyncedFileLocation.INTERNAL,
					connectionSettings = StoredMediaCenterConnectionSettings()
				).toPromise()
			},
            fakePrivateDirectoryLookup,
			fakePrivateDirectoryLookup
		)
		syncDirectoryLookup.promiseSyncDirectory(LibraryId(1)).toExpiringFuture().get()
	}

	@Test
	fun `then the drive is the one with the most space`() {
		assertThat(file!!.path).isEqualTo("/storage/0/my-private-sd-card/1")
	}
}
